pkg/asset/store.go

```diff
@@ -250,86 +222,125 @@ func (s *StoreImpl) fetch(asset Asset, indent string) error {
 	return nil
 }
 
-func (s *StoreImpl) loadFromOnDisk(asset Asset, indent string) (*assetState, error) {
+// load loads the asset and all of its ancestors from on-disk and the state file.
+func (s *StoreImpl) load(asset Asset, indent string) (*assetState, error) {
+	logrus.Debugf("%sLoading %q...", indent, asset.Name())
+
 	// Stop descent if the asset has already been loaded.
 	if state, ok := s.assets[reflect.TypeOf(asset)]; ok {
 		return state, nil
 	}
-	anyParentsDirty := false
+
 	// Load dependencies from on-disk.
+	anyParentsDirty := false
 	for _, d := range asset.Dependencies() {
-		state, err := s.loadFromOnDisk(d, increaseIndent(indent))
+		state, err := s.load(d, increaseIndent(indent))
 		if err != nil {
 			return nil, err
 		}
-		if state.anyParentsDirty || (state.presentOnDisk && !state.onDiskMatchesStateFile) {
+		if state.anyParentsDirty || state.source == onDiskSource {
 			anyParentsDirty = true
 		}
 	}
-	// Try to load from the provided files.
-	foundOnDisk := false
-	as, isWritable := asset.(WritableAsset)
+
+	// Try to load from on-disk.
+	var (
+		onDiskAsset WritableAsset
+		foundOnDisk bool
+	)
+	_, isWritable := asset.(WritableAsset)
 	if isWritable {
-		logrus.Debugf("%sLoading %q from on-disk...", indent, asset.Name())
+		onDiskAsset = reflect.New(reflect.TypeOf(asset).Elem()).Interface().(WritableAsset)
 		var err error
-		foundOnDisk, err = as.Load(s.fileFetcher)
+		foundOnDisk, err = onDiskAsset.Load(s.fileFetcher)
 		if err != nil {
 			return nil, errors.Wrapf(err, "failed to load asset %q", asset.Name())
 		}
 	}
-	if foundOnDisk {
-		logrus.Debugf("%sFound %q on-disk", indent, asset.Name())
-	}
-	matchesStateFile := false
-	if foundOnDisk && s.isAssetInState(asset) {
-		logrus.Debugf("%sLoading %q from both state file and target directory", indent, asset.Name())
 
-		stateAsset := reflect.New(reflect.TypeOf(asset).Elem()).Interface().(Asset)
-		if err := s.loadAssetFromState(stateAsset); err != nil {
-			return nil, errors.Wrapf(err, "failed to load asset %q from state file", asset.Name())
+	// Try to load from state file.
+	var (
+		stateFileAsset         Asset
+		foundInStateFile       bool
+		onDiskMatchesStateFile bool
+	)
+	// Do not need to bother with loading from state file if any of the parents
+	// are dirty because the asset must be re-generated in this case.
+	if !anyParentsDirty {
```

Comment:
we do we load `on-disk` if parents are dirty? but skip load `from-state-file`

Reply:
Because when the asset is on disk, then we need to purge it later, regardless of whether the parents are dirty. One way or another, we need to find the on-disk files for the asset to delete them.